File: examples/generate-signed-url.js
// Example: Generate Presigned URL for Garage S3
// Usage: node generate-signed-url.js <bucket> <key>

import { S3Client, GetObjectCommand } from "@aws-sdk/client-s3";
import { getSignedUrl } from "@aws-sdk/s3-request-presigner";

// Check args
const bucket = process.argv[2];
const key = process.argv[3];

if (!bucket || !key) {
  console.error("Usage: node generate-signed-url.js <bucket> <key>");
  process.exit(1);
}

// Configuration (Load from env or default)
const endpoint = process.env.S3_ENDPOINT || "http://localhost:3900";
const region = process.env.S3_REGION || "garage";
const accessKeyId = process.env.AWS_ACCESS_KEY_ID;
const secretAccessKey = process.env.AWS_SECRET_ACCESS_KEY;

if (!accessKeyId || !secretAccessKey) {
  console.error("Error: AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY environment variables must be set.");
  process.exit(1);
}

const s3 = new S3Client({
  endpoint,
  region,
  credentials: {
    accessKeyId,
    secretAccessKey
  },
  forcePathStyle: true // Garage/MinIO often require this for http://localhost
});

async function generateUrl() {
  try {
    const command = new GetObjectCommand({
      Bucket: bucket,
      Key: key
    });

    // URL valid for 1 hour (3600 seconds)
    const url = await getSignedUrl(s3, command, { expiresIn: 3600 });

    console.log("✅ Signed URL generated successfully (valid for 1 hour):");
    console.log("-------------------------------------------------------");
    console.log(url);
    console.log("-------------------------------------------------------");
  } catch (err) {
    console.error("Error generating signed URL:", err);
  }
}

generateUrl();

File: examples/generate-signed-url.py
import boto3
import sys
import os
from botocore.client import Config

# Check args
if len(sys.argv) < 3:
    print("Usage: python generate-signed-url.py <bucket> <key>")
    sys.exit(1)

bucket = sys.argv[1]
key = sys.argv[2]

# Configuration
endpoint = os.getenv('S3_ENDPOINT', 'http://localhost:3900')
access_key = os.getenv('AWS_ACCESS_KEY_ID')
secret_key = os.getenv('AWS_SECRET_ACCESS_KEY')
region = os.getenv('S3_REGION', 'garage')

if not access_key or not secret_key:
    print("Error: AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY environment variables must be set.")
    sys.exit(1)

# Configure S3 client
s3 = boto3.client('s3',
    endpoint_url=endpoint,
    aws_access_key_id=access_key,
    aws_secret_access_key=secret_key,
    config=Config(signature_version='s3v4'),
    region_name=region
)

try:
    url = s3.generate_presigned_url(
        ClientMethod='get_object',
        Params={
            'Bucket': bucket,
            'Key': key
        },
        ExpiresIn=3600
    )
    
    print("✅ Signed URL generated successfully (valid for 1 hour):")
    print("-------------------------------------------------------")
    print(url)
    print("-------------------------------------------------------")

except Exception as e:
    print(f"Error: {e}")
